R/data.table.R

```diff
@@ -1122,8 +1122,14 @@ replace_dot_alias = function(e) {
           if (is.name(lhs)) {
             lhs = as.character(lhs)
           } else {
-            # e.g. (MyVar):= or get("MyVar"):=
-            lhs = eval(lhs, parent.frame(), parent.frame())
+            #i.e lhs is names(.SD) || setdiff(names(.SD), cols) || (cols)
+            replace_names_sd = function(e, cols){
+              if (length(e) == 1L) return(e)
+              if (e[[1L]] == quote(names) && is.name(e[[2L]]) && e[[2L]] == quote(.SD)) return(cols)
```

Comment:
I think this will cause trouble if it's a namespaced call in LHS. We don't need to support it, but should be aware of the possibility and test:

```r
DT[, base::names(.SD) := ...]
```

```suggestion
              if (e %iscall% "names" && is.name(e2 <- e[[2L]]) && e2 == ".SD") return(cols)
```

Reply:
If we wanted to support namespaced calls like `base::names(.SD)`, it seems like doing it from `%iscall%` might be OK. I've seen you guys already specifically address this issue elsewhere, like `data.table::first()`, but it might be handy.

This is only partially tested - I peeked at what you did to fix data.table::first() and the like.

Regardless, I'm including your suggested changes. Thank you.

``` r
"%iscall2%" = function(e, f) {
   ans = FALSE
   if (is.call(e)) {
     test_exp = e[[1L]]

     test_exp_name = if (is.symbol(test_exp)) test_exp else test_exp[[3L]]
     ans = test_exp_name %chin% f
   }
   return(ans)
 }
```

Reply:
I retract my suggested change to `iscall`. A simpler way would be to do `is.call(q[[1L]]) && deparse(q[[1L]]) %chin% c('names', 'base::names') && ...`  if we wanted to support namespaced calls. My commit will still just be your suggestion for now.